[DOC] Topic Operator topic store (#4415)
strimzi/strimzi-kafka-operator: documentation/modules/operators/con-topic-operator-store.adoc

// Module included in the following assemblies:
//
// assembly-using-the-topic-operator.adoc

[id='ref-topic-operator-store-{context}']
= Topic Operator topic store

[role="_abstract"]

> The Topic Operator uses persistent storage to store topic metadata describing topic configuration as key-value pairs.
> The _topic store_ is based on the Kafka Streams key-value mechanism, which uses persistent storage with local and remote components.

Reviewer: The term `persistent storage` is misleading here since the topic operator does not use persistent storage. Maybe this should say something like this? Also, no clue what _persistent storage with local and remote components_ means.

```suggestion
The Topic Operator uses Kafka to store topic metadata describing topic configuration as key-value pairs.
The _topic store_ is based on the Kafka Streams key-value mechanism, which uses Kafka topics to persist the state.
```

Author: Updated as suggested. I was trying to convey the two aspects: in-memory and backup

> Topic metadata is stored and accessed locally in-memory.
> Updates from operations applied to the local in-memory topic store are persisted to a backup topic store on disk.

Reviewer: This is again confusing since the previous paragraph just explained how it uses persistent storage and here it suddenly said it is in-memory. Maybe this would be better?

```suggestion
Topic metadata are cached in-memory and accessed locally.
Updates from operations applied to the local in-memory cache are persisted to a backup topic store on disk.
```

Author: Thanks. _Cache_ helps here.

The topic store is continually synchronized with updates from Kafka topics or Kubernetes `KafkaTopic` custom resources.
Operations are handled rapidly with the topic store set up this way,

> but should the in-memory topic store crash it is automatically repopulated from the persistent storage.

Reviewer: ```suggestion
but should the in-memory cache crash it is automatically repopulated from the persistent storage.
```

== Internal topic store topics

Internal topics support the handling of topic metadata.

`__strimzi_store_topic`:: Input topic for storing the topic metadata

> `__strimzi-topic-operator-kstreams-topic-store-changelog`:: Retains a log of compacted topic store values

Reviewer: Can we add something like:

```
IMPORTANT: Do not delete these topics. Their deletion can cause Topic Operator failures.
```

Author: IMPORTANT: Do not delete these topics, as they are essential to the running of the Topic Operator.

== Migrating topic metadata from ZooKeeper

In previous releases of Strimzi, topic metadata was stored in ZooKeeper.
The new process removes this requirement, bringing the metadata into the Kafka cluster, and under the control of the Topic Operator.

When upgrading to Strimzi {ProductVersion}, the transition to Topic Operator control of the topic store is seamless.

> Metadata is found and migrated from ZooKeeper, and the old store is cleansed.

Reviewer: Is the old store deleted? Or just cleansed? I think that for lot of Kafka users this might mean two different things.

Author: @alesj - Ales -- when upgrading is the old store just cleansed or deleted?

Reviewer 2: Is erased a better verb?

Reviewer 3: ```                    log.info("Deleting ZK topics path: {}", topicsPath);
                    zk.delete(topicsPath, -1);
```

Reviewer 3: The ZK node holding the topcs data is deleted ...

Reviewer: @PaulRMellor So I guess we can just say _deleted_?

Author: Updated

== Downgrading to a Strimzi version that uses ZooKeeper to store topic metadata

If you are reverting back to a version of Strimzi that uses ZooKeeper for the storage of topic metadata,
you downgrade your Cluster Operator to the previous version,
then downgrade Kafka brokers and client applications to the previous Kafka version as standard.

Additionally, you must delete the topics created for the topic store using an `exec` command, specifying the bootstrap address of the Kafka cluster and the name of a broker pod.
For example, where `my-cluster-kafka-0` is the name of the broker pod:

[source,shell]
----

> kubectl exec -ti my-cluster-kafka-0 -- ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --topic __strimzi-topic-operator-kstreams-topic-store-changelog --delete && ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --topic __strimzi_store_topic --delete

Reviewer: Thsi should use its own pod as in the command I suggested above.

Author: I've replaced with the kube-admin command, as suggested above

----

The Topic Operator will reconstruct the Zookeeper topic metadata from the state of the topics in Kafka.

> NOTE: When scaling a Kafka cluster, the topic store topics must also be deleted so that partitions can be reassigned.

Reviewer: I don't think this is true. Why would we delete the topic just because of scaling?

Author: @ales - I'm dropping this note

== Scaling and topics

> The recommended configuration for topics is to have a topic replication factor of 3, so that there are 2 in-sync replicas on other brokers.

Reviewer: ```suggestion
The recommended configuration for the Topic Operator topics is to have a topic replication factor of 3 and minimum of 2 in-sync replicas.
```

> If a broker is unavailable, at least one other broker is in-sync.

Reviewer: ```suggestion
```

[source,shell,subs="+quotes,attributes"]
----
apiVersion: {KafkaTopicApiVersion}
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:

>   partitions: 1

Reviewer: Is 1 partition just an example? Or is it a requirement? Can the TO work with multiple partitions @alesj?

Reviewer 3: Hmmm, good question, since this is the topic controlled by Kafka Streams -- this is the input topic into our topology.
@tombentley ^^ how with the topics controlled by Kafka Streams?

Reviewer 3: OK, afais, it goes like this:
* you (can) configure number of Streams threads

```   
public static final String NUM_STREAM_THREADS_CONFIG = "num.stream.threads";
private static final String NUM_STREAM_THREADS_DOC = "The number of threads to execute stream processing.";
```

and then those input topic partitions get distributed among all cluster threads / consumers.

Reviewer 4: I can't think of a reason why it wouldn't work with >1 partition, but nor do I see that >1 would be needed.

Author: I can add another callout:

> <1> The number of partitions for the topic. Generally, 1 partition is sufficient.

  replicas: 3 <1>
  config:
    min.insync.replicas=2 <2>
  #...
----
<1> The number of replica topics. Currently, this cannot be changed in the `KafkaTopic` resource, but it can be changed using the `kafka-reassign-partitions.sh` tool.
<2> The minimum number of replica partitions that a message must be successfully written to, or an exception is raised.

NOTE: In-sync replicas are used in conjunction with the `acks` configuration for a producer application.

> The `acks` configuration determines the number of follower partitions a message must be replicated to before the message is acknowledged as successfully received.

Reviewer: This is true. But what does it mean for the user? We should either add how they can configure it for the TO. Or we should just say that the TO always runs with `acks=all`.

Author: Not sure if I'm capturing what you think we need in this update:

> NOTE: In-sync replicas are used in conjunction with the `acks` configuration for a producer application.
> The `acks` configuration determines the number of follower partitions a message must be replicated to before the message is acknowledged as successfully received. The Topic Operator runs with `acks=all` by default, whereby messages must be acknowledged by all in-sync replicas.

Reviewer: `runs with acks=all by default` ... I think it is not possible to change it. So I would leave out the `by default` maybe?

Author: Okay.

> When scaling Kafka clusters by adding or removing brokers, replicas are not reassigned automatically.

Reviewer: ```suggestion
When scaling Kafka clusters by adding or removing brokers, replication factor configuration is not changed and replicas are not reassigned automatically.
```

> However, you can use the `kafka-reassign-partitions.sh` tool to change the replication factor, and manually reassign replicas to brokers.

Reviewer: Should we add link to https://strimzi.io/docs/operators/latest/full/using.html#con-partition-reassignment-str here?

Author: I've put links to all relevant sections at the end , including this one

> Alternatively, though Cruise Control cannot change the replication factor for topics,

Reviewer: I assume Cruise Control can do this, but the Strimzi - CruiseControl integration cannot. Maybe we should make it more clear to not mislead anyone about Cruise Control itself.

Author: Updated:

> Alternatively, though the integration of Cruise Control for Strimzi cannot change the replication factor for topics,
> the optimization proposals it generates for rebalancing Kafka include commands that transfer partition replicas and change partition leadership.

// Module included in the following assemblies:
//
// assembly-using-the-topic-operator.adoc

[id='ref-topic-operator-store-{context}']
= Topic Operator topic store

[role="_abstract"]
The Topic Operator uses Kafka to store topic metadata describing topic configuration as key-value pairs.
The _topic store_ is based on the Kafka Streams key-value mechanism, which uses Kafka topics to persist the state.

Topic metadata is cached in-memory and accessed locally.
Updates from operations applied to the local in-memory cache are persisted to a backup topic store on disk.
The topic store is continually synchronized with updates from Kafka topics or Kubernetes `KafkaTopic` custom resources.
Operations are handled rapidly with the topic store set up this way,
but should the in-memory cache crash it is automatically repopulated from the persistent storage.

== Internal topic store topics

> Internal topics support the handling of topic metadata.

Reviewer 2: ```suggestion
Internal topics support the handling of topic metadata stored in the topic store.
```

`__strimzi_store_topic`:: Input topic for storing the topic metadata
`__strimzi-topic-operator-kstreams-topic-store-changelog`:: Retains a log of compacted topic store values

> IMPORTANT: Do not delete these topics, as they are essential to the running of the Topic Operator.

Reviewer 2: Maybe use a warning here.

== Migrating topic metadata from ZooKeeper

> In previous releases of Strimzi, topic metadata was stored in ZooKeeper.

Reviewer 2: This information might be helpful earlier on as well.

Author: If we keep the mention only here, we can drop this section cleanly when it is no longer needed.

The new process removes this requirement, bringing the metadata into the Kafka cluster, and under the control of the Topic Operator.

When upgrading to Strimzi {ProductVersion}, the transition to Topic Operator control of the topic store is seamless.
Metadata is found and migrated from ZooKeeper, and the old store is cleansed.

== Downgrading to a Strimzi version that uses ZooKeeper to store topic metadata

If you are reverting back to a version of Strimzi earlier than 0.22, which uses ZooKeeper for the storage of topic metadata,
you still downgrade your Cluster Operator to the previous version,
then downgrade Kafka brokers and client applications to the previous Kafka version as standard.

> However, you must also delete the topics created for the topic store using a `kafka-admin` command, specifying the bootstrap address of the Kafka cluster.

Reviewer 2: ```suggestion
However, you must also delete the topics that were created for the topic store using a `kafka-admin` command, specifying the bootstrap address of the Kafka cluster.
```

For example:

[source,shell]
----
kubectl run kafka-admin -ti --image={DockerKafkaImageCurrent} --rm=true --restart=Never -- ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --topic __strimzi-topic-operator-kstreams-topic-store-changelog --delete && ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --topic __strimzi_store_topic --delete
----

The command must correspond to the type of listener and authentication used to access the Kafka cluster.

The Topic Operator will reconstruct the ZooKeeper topic metadata from the state of the topics in Kafka.

== Topic Operator topic replication and scaling

> The recommended configuration for the Topic Operator topics is to have a topic replication factor of 3, and a minimum of 2 in-sync replicas.

Reviewer 2: ```suggestion
The recommended configuration for topics managed by the Topic Operator is a topic replication factor of 3, and a minimum of 2 in-sync replicas.
```

[source,shell,subs="+quotes,attributes"]
----
apiVersion: {KafkaTopicApiVersion}
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 1
  replicas: 3 <1>
  config:
    min.insync.replicas=2 <2>
  #...
----

> <1> The number of replica topics. Currently, this cannot be changed in the `KafkaTopic` resource, but it can be changed using the `kafka-reassign-partitions.sh` tool.

Reviewer 2: ```suggestion
<1> The number of partition replicas. Currently, this cannot be changed in the `KafkaTopic` resource, but it can be changed using the `kafka-reassign-partitions.sh` tool.
```

<2> The minimum number of replica partitions that a message must be successfully written to, or an exception is raised.

> NOTE: In-sync replicas are used in conjunction with the `acks` configuration for a producer application.

Reviewer 2: ```suggestion
NOTE: In-sync replicas are used in conjunction with the `acks` configuration for producer applications.
```

The `acks` configuration determines the number of follower partitions a message must be replicated to before the message is acknowledged as successfully received.
The Topic Operator runs with `acks=all`, whereby messages must be acknowledged by all in-sync replicas.

When scaling Kafka clusters by adding or removing brokers, replication factor configuration is not changed and replicas are not reassigned automatically.
However, you can use the `kafka-reassign-partitions.sh` tool to change the replication factor, and manually reassign replicas to brokers.

Alternatively, though the integration of Cruise Control for Strimzi cannot change the replication factor for topics,
the optimization proposals it generates for rebalancing Kafka include commands that transfer partition replicas and change partition leadership.

== Handling changes to topics

> A fundamental problem that the Topic Operator has to solve is that there is no single source of truth:

Reviewer 2: ```suggestion
A fundamental problem that the Topic Operator needs to solve is that there is no single source of truth:
```

> Both the `KafkaTopic` resource and the Kafka topic can be modified independently of the Topic Operator.

Reviewer 2: ```suggestion
both the `KafkaTopic` resource and the Kafka topic can be modified independently of the Topic Operator.
```

For example, when the Topic Operator is down.

To resolve this, the Topic Operator maintains information about each topic in the topic store.
When a change happens in the Kafka cluster or Kubernetes, it looks at both the state of the other system and the topic store in order to determine what needs to change to keep everything in sync.
The same thing happens whenever the Topic Operator starts, and periodically while it is running.

For example, suppose the Topic Operator is not running, and a `KafkaTopic` called _my-topic_ is created.
When the Topic Operator starts, the topic store does not contain information on my-topic, so it can infer that the `KafkaTopic` was created after it was last running.
The Topic Operator creates the topic corresponding to my-topic, and also stores metadata for my-topic in the topic store.

If you update Kafka topic configuration or apply a change through the `KafkaTopic` custom resource,
the topic store is updated after the Kafka cluster is reconciled.

The topic store also allows the Topic Operator to manage scenarios where the topic configuration is changed in Kafka topics _and_ updated through Kubernetes `KafkaTopic` custom resources,

> so long as the changes are not incompatible.

Reviewer 2: ```suggestion
as long as the changes are not incompatible.
```

For example, it is possible to make changes to the same topic config key, but to different values.
For incompatible changes, the Kafka configuration takes priority, and the `KafkaTopic` is updated accordingly.

NOTE: You can also use the `KafkaTopic` resource to delete topics using a `kubectl delete -f _KAFKA-TOPIC-CONFIG-FILE_` command.
To be able to do this, `delete.topic.enable` must be set to `true` (default) in the `spec.kafka.config` of the Kafka resource.

> 

Reviewer 2: Extra space.
